### PR TITLE
:sparkles: feat: MetaMask setup in Cypress

### PR DESCRIPTION
**Pull request by localtechboy**
## Motivation and context

Install MetaMask in Cypress.

## Does it fix any issue?

https://linear.app/synpress/issue/SYN-124/setup-metamask-extension-for-cypress

## Quality checklist

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough e2e tests.

**⚠️👆 Delete any section you see irrelevant before submitting the pull request 👆⚠️**

### CODE REVIEW

**Comment by 0xSero** [comment]
Few things to note here:

Since we are doing some cacheless setup from what I can tell in your PR, and we're doing the same thing for Playwright in my current WIP PR. It might make sense to create a shared folder of functions for cypress/playwright.

Also have you run into any problems trying to run things using Cache in Cypress?

**Comment by 0xSero** [comment]
You forgot to git ignore the download directory

**Comment by 0xSero** [comment]
1. Unit tests are failing
2. Does this include using cache for Cypress? [Here is a potential setup to make this happen](https://github.com/cypress-io/cypress/issues/18308#issuecomment-2183309755)
3. I'm concerned we're going to have a difficult time with merging and dealing with conflicts given we've been developing in silos for the last month.

I've created a dev branch for v4 beta, I would appreciate if you could merge into there as we prep for V4 beta release, since we've been working on structural changes, leaving prs open for weeks, and dealing with conflicts can get very time consuming.

https://github.com/Synthetixio/synpress/pull/1175

**Comment by localtechboy** [comment]
> 1. Unit tests are failing
> 2. Does this include using cache for Cypress? [Here is a potential setup to make this happen](https://github.com/cypress-io/cypress/issues/18308#issuecomment-2183309755)
> 3. I'm concerned we're going to have a difficult time with merging and dealing with conflicts given we've been developing in silos for the last month.
> 
> I've created a dev branch for v4 beta, I would appreciate if you could merge into there as we prep for V4 beta release, since we've been working on structural changes, leaving prs open for weeks, and dealing with conflicts can get very time consuming.
> 
> #1175

1. Yes, I pushed a few fixes today.
2. No cache. It will be a separated task to cover that.

How is it better to use different branch? There will be the same conflicts here and there right?

**Comment by 0xSero** [comment]
> How is it better to use different branch? There will be the same conflicts here and there right?

The reason I am pushing for us using the new-dawn-dev branch is so that we don't have to hold off on merging our work together, for example:

1. Right now we can't merge to ```new-dawn``` without thorough reviews, which is causing us to work in separate fragmented PRs, only dealing and updating PRs as ones make it through review
2. When we make big structural changes we need to update/deploy our packages to npm, however some changes need to be done using ```workspace:*``` since our repo has many interdependencies and as far as I know only @drptbl can update npm
3. it will allow us to work closer together and consider each-other's work, to hopefully reduce wasteful merge conflicts.
4. It will allow us to organize the v4 BETA release under 1 branch and do a thorough test prior to launch.

I know it's a bit of a pace change but I'd really appreciate if we can adopt this.

File: .gitignore [hunk 1]
```diff
@@ -37,6 +37,9 @@ test-results
 playwright-report
 playwright/.cache
 
+### Cypress
```
Comment (0xSero) [inline .gitignore:40]: Could we use **/downloads instead? Since we're going to need to use this pattern elsewhere

File: wallets/metamask/src/cypress/MetaMask.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { lockPage } from '../selectors'
+import { MetaMaskAbstract } from '../type/MetaMaskAbstract'
+
+// @ts-ignore
```
Comment (0xSero) [inline wallets/metamask/src/cypress/MetaMask.ts:4]: can we refrain from using ```ts-ignore``` and ```any```

File: wallets/metamask/src/prepareExtension.ts [hunk 3]
```diff
@@ -1,14 +1,25 @@
...
 export const EXTENSION_DOWNLOAD_URL = `https://github.com/MetaMask/metamask-extension/releases/download/v${DEFAULT_METAMASK_VERSION}/metamask-chrome-${DEFAULT_METAMASK_VERSION}.zip`
 
-export async function prepareExtension() {
-  const cacheDirPath = ensureCacheDirExists()
+export async function prepareExtension(forceCache = true) {
+  let outputDir = ''
+  if (forceCache) {
+    outputDir = ensureCacheDirExists()
+  } else {
+    outputDir = path.resolve('./', 'downloads')
```
Comment (0xSero) [inline wallets/metamask/src/prepareExtension.ts]: can we ensure we use platform agnostic path resolutions? 

Based on the patterns from #1175 we can now use Synpress directly on Windows, this wouldn't be compatible. Check the linked PR on how to do that.

File: wallets/metamask/src/type/MetaMaskAbstract.ts [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+   * @param password - The password of the MetaMask wallet.
+   * @param extensionId - The extension ID of the MetaMask extension. Optional if no interaction with the dapp is required.
+   *
+   * @returns A new instance of the MetaMask class.
+   */
+  constructor(
+    /**
+     * The password of the MetaMask wallet.
+     */
+    readonly password: string,
```
[the remark this answers is not in the document]
Comment (0xSero) [reply wallets/metamask/src/type/MetaMaskAbstract.ts:16]: Please resolve this

File: wallets/metamask/test/cypress/metamask.cy.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+it('should connect to ethereum', () => {
+  cy.window().then((window) => {
+    // @ts-ignore
```
Comment (0xSero) [inline wallets/metamask/test/cypress/metamask.cy.ts:3]: refrain from using ts-ignore

**Review by 0xSero — CHANGES_REQUESTED** [review]
I am not sure if there's a way for us to shift to using better relative imports, the ```../../``` pattern is pretty ugly and confusing, but it's not critical

**Comment by localtechboy** [comment]
> > How is it better to use different branch? There will be the same conflicts here and there right?
> 
> The reason I am pushing for us using the new-dawn-dev branch is so that we don't have to hold off on merging our work together, for example:
> 
> 1. Right now we can't merge to `new-dawn` without thorough reviews, which is causing us to work in separate fragmented PRs, only dealing and updating PRs as ones make it through review
> 2. When we make big structural changes we need to update/deploy our packages to npm, however some changes need to be done using `workspace:*` since our repo has many interdependencies and as far as I know only @drptbl can update npm
> 3. it will allow us to work closer together and consider each-other's work, to hopefully reduce wasteful merge conflicts.
> 4. It will allow us to organize the v4 BETA release under 1 branch and do a thorough test prior to launch.
> 
> I know it's a bit of a pace change but I'd really appreciate if we can adopt this.

But how the new branch will resolve the conflicts?

File: wallets/metamask/src/cypress/MetaMask.ts (hunk 2 above)
Comment (localtechboy) [reply wallets/metamask/src/cypress/MetaMask.ts:4]: it will be included in a separated PR as mentioned in the comment right below

**Comment by 0xSero** [comment]
> > > How is it better to use different branch? There will be the same conflicts here and there right?
> > 
> > 
> > The reason I am pushing for us using the new-dawn-dev branch is so that we don't have to hold off on merging our work together, for example:
> > 
> > 1. Right now we can't merge to `new-dawn` without thorough reviews, which is causing us to work in separate fragmented PRs, only dealing and updating PRs as ones make it through review
> > 2. When we make big structural changes we need to update/deploy our packages to npm, however some changes need to be done using `workspace:*` since our repo has many interdependencies and as far as I know only @drptbl can update npm
> > 3. it will allow us to work closer together and consider each-other's work, to hopefully reduce wasteful merge conflicts.
> > 4. It will allow us to organize the v4 BETA release under 1 branch and do a thorough test prior to launch.
> > 
> > I know it's a bit of a pace change but I'd really appreciate if we can adopt this.
> 
> But how the new branch will resolve the conflicts?

I explained it ^

TLDR: 

1. faster merges because less review pressure
2. because faster merges we can work off the latest changes

[It's like having a dev branch](https://stackoverflow.com/questions/32826370/why-do-we-need-dev-branch), have you ever used a dev branch? Because you're not the only one pushing code now we need to find a way to keep in sync.

**Comment by localtechboy** [comment]
> > > > How is it better to use different branch? There will be the same conflicts here and there right?
> > > 
> > > 
> > > The reason I am pushing for us using the new-dawn-dev branch is so that we don't have to hold off on merging our work together, for example:
> > > 
> > > 1. Right now we can't merge to `new-dawn` without thorough reviews, which is causing us to work in separate fragmented PRs, only dealing and updating PRs as ones make it through review
> > > 2. When we make big structural changes we need to update/deploy our packages to npm, however some changes need to be done using `workspace:*` since our repo has many interdependencies and as far as I know only @drptbl can update npm
> > > 3. it will allow us to work closer together and consider each-other's work, to hopefully reduce wasteful merge conflicts.
> > > 4. It will allow us to organize the v4 BETA release under 1 branch and do a thorough test prior to launch.
> > > 
> > > I know it's a bit of a pace change but I'd really appreciate if we can adopt this.
> > 
> > 
> > But how the new branch will resolve the conflicts?
> 
> I explained it ^
> 
> TLDR:
> 
> 1. faster merges because less review pressure
> 2. because faster merges we can work off the latest changes
> 
> [It's like having a dev branch](https://stackoverflow.com/questions/32826370/why-do-we-need-dev-branch), have you ever used a dev branch? Because you're not the only one pushing code now we need to find a way to keep in sync.

But it's not helping with the conflicts. It's not a solution if we will have another, not synchronised branch. Lets try to make the way we have better aligned :)

**Comment by 0xSero** [comment]
> > > > > How is it better to use different branch? There will be the same conflicts here and there right?
> > > > 
> > > > 
> > > > The reason I am pushing for us using the new-dawn-dev branch is so that we don't have to hold off on merging our work together, for example:
> > > > 
> > > > 1. Right now we can't merge to `new-dawn` without thorough reviews, which is causing us to work in separate fragmented PRs, only dealing and updating PRs as ones make it through review
> > > > 2. When we make big structural changes we need to update/deploy our packages to npm, however some changes need to be done using `workspace:*` since our repo has many interdependencies and as far as I know only @drptbl can update npm
> > > > 3. it will allow us to work closer together and consider each-other's work, to hopefully reduce wasteful merge conflicts.
> > > > 4. It will allow us to organize the v4 BETA release under 1 branch and do a thorough test prior to launch.
> > > > 
> > > > I know it's a bit of a pace change but I'd really appreciate if we can adopt this.
> > > 
> > > 
> > > But how the new branch will resolve the conflicts?
> > 
> > 
> > I explained it ^
> > TLDR:
> > 
> > 1. faster merges because less review pressure
> > 2. because faster merges we can work off the latest changes
> > 
> > [It's like having a dev branch](https://stackoverflow.com/questions/32826370/why-do-we-need-dev-branch), have you ever used a dev branch? Because you're not the only one pushing code now we need to find a way to keep in sync.
> 
> But it's not helping with the conflicts. It's not a solution if we will have another, not synchronised branch. Lets try to make the way we have better aligned :)

You're being pedantic, dev branches are a well defined, useful standard that is popular and widely adopted for a reason.
The way you're working makes more work for me, and vice-versa.

For example, I already got things working on Windows, so if you're working off the latest branch you can just adopt the solution so I don't have to jump in and clean after you, or get my work blocked for another x time because windows works in 1 wallet and not another.

There are people using new-dawn  ```0.0.1-alpha.7```, progress is made much slower there, which is why we are not in sync, having a dev branch like everyone else is a clean and simple solution. (:

**Comment by localtechboy** [comment]
> > > > > > How is it better to use different branch? There will be the same conflicts here and there right?
> > > > > 
> > > > > 
> > > > > The reason I am pushing for us using the new-dawn-dev branch is so that we don't have to hold off on merging our work together, for example:
> > > > > 
> > > > > 1. Right now we can't merge to `new-dawn` without thorough reviews, which is causing us to work in separate fragmented PRs, only dealing and updating PRs as ones make it through review
> > > > > 2. When we make big structural changes we need to update/deploy our packages to npm, however some changes need to be done using `workspace:*` since our repo has many interdependencies and as far as I know only @drptbl can update npm
> > > > > 3. it will allow us to work closer together and consider each-other's work, to hopefully reduce wasteful merge conflicts.
> > > > > 4. It will allow us to organize the v4 BETA release under 1 branch and do a thorough test prior to launch.
> > > > > 
> > > > > I know it's a bit of a pace change but I'd really appreciate if we can adopt this.
> > > > 
> > > > 
> > > > But how the new branch will resolve the conflicts?
> > > 
> > > 
> > > I explained it ^
> > > TLDR:
> > > 
> > > 1. faster merges because less review pressure
> > > 2. because faster merges we can work off the latest changes
> > > 
> > > [It's like having a dev branch](https://stackoverflow.com/questions/32826370/why-do-we-need-dev-branch), have you ever used a dev branch? Because you're not the only one pushing code now we need to find a way to keep in sync.
> > 
> > 
> > But it's not helping with the conflicts. It's not a solution if we will have another, not synchronised branch. Lets try to make the way we have better aligned :)
> 
> You're being pedantic, dev branches are a well defined, useful standard that is popular and widely adopted for a reason. The way you're working makes more work for me, and vice-versa.
> 
> For example, I already got things working on Windows, so if you're working off the latest branch you can just adopt the solution so I don't have to jump in and clean after you, or get my work blocked for another x time because windows works in 1 wallet and not another.
> 
> There are people using new-dawn `0.0.1-alpha.7`, progress is made much slower there, which is why we are not in sync, having a dev branch like everyone else is a clean and simple solution. (:

I am not saying about dev branch approach but about conflicts. You said you want to avoid them. I am just saying another branch will not resolve any conflict.

Furthermore how merging this branch somewhere else will help you will the conflicts in the future?

File: .gitignore [hunk 6]
```diff
@@ -37,6 +37,9 @@ test-results
...
+wallets/metamask/downloads
```
Comment (drptbl) [inline .gitignore]: use `wallets/**/downloads`

File: wallets/metamask/cypress.config.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+export default defineConfig({
+  chromeWebSecurity: false,
+  e2e: {
+    baseUrl: 'http://localhost:9999',
+    specPattern: 'test/cypress/**/*.cy.{js,jsx,ts,tsx}',
+    supportFile: 'src/cypress/support/e2e.{js,jsx,ts,tsx}',
+    testIsolation: false,
+    async setupNodeEvents(on, config) {
+      return installSynpress(on, config)
```
Comment (drptbl) [inline wallets/metamask/cypress.config.ts]: is `installSynpress` a good name here?
maybe `configureSynpress` would be better?

File: wallets/metamask/package.json [hunk 8]
```diff
@@ -15,15 +21,16 @@
   ],
   "scripts": {
     "build": "pnpm run clean && pnpm run build:dist && pnpm run build:types",
-    "build:cache": "synpress-cache test/wallet-setup",
-    "build:cache:headless": "synpress-cache test/wallet-setup --headless",
-    "build:cache:headless:force": "synpress-cache test/wallet-setup --headless --force",
+    "build:cache": "synpress-cache test/playwright/wallet-setup",
+    "build:cache:headless": "synpress-cache test/playwright/wallet-setup --headless",
```
Comment (drptbl) [inline wallets/metamask/package.json:25]: should we also adjust npm run script names and add `:playwright` at the end like we do with cypress?

File: wallets/metamask/src/prepareExtension.ts (hunk 3 above)
Comment (drptbl) [reply wallets/metamask/src/prepareExtension.ts]: @matstyler worth fixing?
example:
https://github.com/Synthetixio/synpress/pull/1175/files#diff-771d1fb7e6025d048bb38f171dfa5fb00f97044447ec1e5c2495312e283a9a0aR13-R14

**Review by drptbl — CHANGES_REQUESTED** [review]

File: wallets/metamask/cypress.config.ts (hunk 7 above)
Comment (localtechboy) [reply wallets/metamask/cypress.config.ts]: what do you think about "configureBeforeSynpress"?

File: wallets/metamask/package.json [hunk 9]
```diff
@@ -15,17 +21,18 @@
   ],
   "scripts": {
     "build": "pnpm run clean && pnpm run build:dist && pnpm run build:types",
-    "build:cache": "synpress-cache test/wallet-setup",
-    "build:cache:headless": "synpress-cache test/wallet-setup --headless",
-    "build:cache:headless:force": "synpress-cache test/wallet-setup --headless --force",
+    "build:cache": "synpress-cache test/playwright/wallet-setup",
```
Comment (drptbl) [inline wallets/metamask/package.json:24]: `build:playwright:cache`?

File: wallets/metamask/src/prepareExtension.ts [hunk 10]
```diff
@@ -1,14 +1,26 @@
...
+    outputDir = process.platform === 'win32' ? `file:\\\\\\${outputDir}` : outputDir
```
Comment (drptbl) [inline wallets/metamask/src/prepareExtension.ts]: why making it complicated, instead of doing
```
    outputDir = process.platform === 'win32' ? `file:\\\\\\${outputDir}` : path.resolve('./', 'downloads')
```
?

File: wallets/ethereum-wallet-mock/cypress.config.ts [hunk 11]
```diff
@@ -10,7 +10,7 @@ export default defineConfig({
     fixturesFolder: 'src/cypress/fixtures',
     testIsolation: false,
     async setupNodeEvents(on, config) {
-      return installSynpress(on, config)
+      return configureBeforeSynpress(on, config)
```
Comment (drptbl) [inline wallets/ethereum-wallet-mock/cypress.config.ts]: `configureBeforeSynpress`?

why not
`configureSynpress`?

**Review by drptbl — CHANGES_REQUESTED** [review]

File: wallets/metamask/src/prepareExtension.ts (hunk 10 above)
Comment (localtechboy) [reply wallets/metamask/src/prepareExtension.ts]: are you sure it will work?

File: wallets/metamask/package.json (hunk 9 above)
Comment (localtechboy) [reply wallets/metamask/package.json:24]: end goal will be to use it for both: Playwright and Cypress

**Review by drptbl — APPROVED** [review]
